### PR TITLE
#107307 Adding additional information to "cargo test --list" output

### LINKED ISSUE #107307: Implementing "<test_binary> --list --format json" for use by IDE test explorers / runners
[consider the default output of ```cargo new a --lib```]

## Issue
Test explorers in editors (e.g. VSCode, see pic below) are unable to navigate to test location in the editor from the test explorer.

## Cause
```cargo test --list -q``` used by the editors to get list of tests do not have the source and line / column info.

## Description
Currently ```cargo test --list -q``` shows the following output
```
tests::it_works: test
```

This is used by editor extensions e.g. [vscode rust test adapter](https://marketplace.visualstudio.com/items?itemName=swellaby.vscode-rust-test-adapter) to show & run tests from the tests pane in VS Code. e.g.
image

While it is possible to run and debug the tests from the test explorer, it is not possible to navigate back to the test location as output above is missing that information. E.g. you cannot double click on a test on the test explorer to navigate to its definition in the editor.

Suggestion is to add the source info to the output of --list as follows
```
tests::it_works: test,<path>|<start_line>|<start_col>|<end_line>|<end_col>
```
## Fix proposal
Proposal for fix (& I only have a plan A, happy to receive further guidance) is to update ```test::TestDescAndFn``` as follows

```rs
// https://github.com/rust-lang/rust/blob/master/library/test/src/types.rs#L118
#[derive(Clone, Debug, PartialEq, Eq, Hash)]
pub struct TestDesc {
    pub name: TestName,
    pub ignore: bool,
    pub ignore_message: Option<&'static str>,
    pub should_panic: options::ShouldPanic,
    pub compile_fail: bool,
    pub no_run: bool,
    pub test_type: TestType,
    pub source_path: &'static str,     // <- Added
    pub start_line: i32,               // <- Added
    pub start_column: i32,             // <- Added
    pub end_line: i32,                 // <- Added
    pub end_column: i32,               // <- Added
}
```
Once this is available, the corresponding changes can be made to [compiler/rustc_builtin_macros/src/test.rs](https://github.com/rust-lang/rust/blob/master/compiler/rustc_builtin_macros/src/test.rs#L251) to emit the above info and then in [library/test/src/console.rs](https://github.com/rust-lang/rust/blob/master/library/test/src/console.rs#L166) to write it out.

### PR DESCRIPTION
**Pull request by parthopdas**
Closes https://github.com/rust-lang/rust/issues/107307

### earlier:
image

### now:
```// test | test_type | ignored_or_active | location_info```
image

### CODE REVIEW

**Comment by parthopdas** [comment]
@petrochenkov please can you help me resolve the build error?

i am able to build by sample projects using the compiler containing the fixes. any idea why these appear?

image

File: library/test/src/tests.rs [hunk 1]
```diff
@@ -63,6 +63,8 @@ fn one_ignored_one_unignored_test() -> Vec<TestDescAndFn> {
                 name: StaticTestName("1"),
                 ignore: true,
                 ignore_message: None,
+                #[cfg(not(bootstrap))]
+                location_info: "src\\lib.rs:10:10: 10:20",
```
Comment (petrochenkov) [inline library/test/src/tests.rs:67]: If these names in tests are never actually used maybe better to set them to empty strings.

File: compiler/rustc_span/src/source_map.rs [hunk 2]
```diff
@@ -467,6 +467,11 @@ impl SourceMap {
         self.span_to_string(sp, FileNameDisplayPreference::Remapped)
     }
 
+    /// Format the span's location information into string
+    pub fn span_to_location_info_string(&self, sp: Span) -> String {
```
Comment (petrochenkov) [inline compiler/rustc_span/src/source_map.rs:471]: Nit: this function is used only once, it's probably better to inline and remove it.

File: .gitignore [hunk 3]
```diff
@@ -21,6 +21,7 @@ Session.vim
 .project
 .favorites.json
 .settings/
+.vs/
```
Comment (petrochenkov) [inline .gitignore:24]: Could you avoid changes to this file?

File: compiler/rustc_builtin_macros/src/test.rs [hunk 4]
```diff
@@ -396,6 +401,12 @@ fn should_ignore_message(cx: &ExtCtxt<'_>, i: &ast::Item) -> Option<Symbol> {
     }
 }
 
+fn item_location_info(cx: &ExtCtxt<'_>, item: &ast::Item) -> Symbol {
+    let ident_sp = cx.with_def_site_ctxt(item.ident.span);
```
Comment (petrochenkov) [inline compiler/rustc_builtin_macros/src/test.rs:405]: I don't think it's necessary to set the def site context here, `span_to_string` doesn't use it.

**Comment by petrochenkov** [comment]
Not sure who is responsible for libtest interfaces like this.
Marked this as waiting on @rust-lang/devtools (it seems like you don't have a dedicated `I-*-nominated` label).

cc @matklad as well, since you may know people who knows something about vscode integration.

**Comment by matklad** [comment]
cc @Veykril 

Some high-level non-normative comments:

- Ideally, IDEs would get this info from static analysis of the code, rather than from compiling and running the binary
- Not everything is an advanced IDE, so having this info available helps low-tech scenarios a lot and, in principle, we should have it
- The current output is de-facto stable, so we can't change it without some extra opt-in flag
- More generally, the standard solution in this space is to emit the info in JSON format, to get us leeway to add new information later in a backwards compatible way
- So we probably want something a-la `--list-tests --message-format json`
- Similarly, I don't know who is the right reviewer here

File: compiler/rustc_span/src/source_map.rs (hunk 2 above)
Comment (parthopdas) [reply compiler/rustc_span/src/source_map.rs:471]: done

File: library/test/src/tests.rs (hunk 1 above)
Comment (parthopdas) [reply library/test/src/tests.rs:67]: done.

File: .gitignore (hunk 3 above)
Comment (parthopdas) [reply .gitignore:24]: this is the visual studio ide creating the file. like .vscode.

with the extension i have just released [(rust-analy](https://marketplace.visualstudio.com/items?itemName=kitamstudios.RustAnalyzer) - for which this PR is, i do expect windows devs to use visual studio more often.

is it okay to retain it?

File: compiler/rustc_builtin_macros/src/test.rs (hunk 4 above)
Comment (parthopdas) [reply compiler/rustc_builtin_macros/src/test.rs:405]: done.

File: library/test/src/console.rs [hunk 5]
```diff
@@ -165,8 +175,11 @@ pub fn list_tests_console(opts: &TestOpts, tests: Vec<TestDescAndFn>) -> io::Res
             }
         };
 
-        writeln!(output, "{name}: {fntype}")?;
-        st.write_log(|| format!("{fntype} {name}\n"))?;
+        #[cfg(bootstrap)]
+        let location_info = "location_info_tbd_during_bootstrap";
+
+        writeln!(output, "{name}: {fntype} | {ignore} | {location_info}")?;
```
[the remark this answers is not in the document]
Comment (parthopdas) [reply library/test/src/console.rs:181]: changed it to more descriptive. ignored | active.

[the remark this answers is not in the document]
Comment (parthopdas) [reply library/test/src/console.rs:181]: we're making it up as we go along - so doing the simplest / most obvious thing. but happy to consider other suggestions.

**Comment by Manishearth** [comment]
Not the devtools team, libtest output is libs or compiler.

**Comment by Manishearth** [comment]
My general take here is that this would require a lot more discussion before landing and probably should not be done as a PR.

And also I think the proper solution is fully machine readable output for cargo test, which requires an RFC.

**Comment by parthopdas** [comment]
@Manishearth some clarifying questions.

> My general take here is that this would require a lot more discussion before landing and probably should not be done as a PR.

are you referring to "--message-format json"? if so i agree. i am happy to start one and then raise the PR.

however if you are referring to this entire change, it will be good to understand your reasoning.

my reasoning is that:
- navigating from test explorer back to code is an important scenario when an engineer has 100s of tests - and this is not possible today. this is typically the case with even simpler of my projects. take e.g. (rust-analyzer.vs)[https://github.com/kitamstudios/rust-analyzer.vs] for which this change is required and which enables rust development on visual studio ide.
- i am just extending the current format <testpath>: <testtype> with additional fields.
- so i feel we should not be holding this up.

> And also I think the proper solution is fully machine readable output for cargo test, which requires an RFC.
i agree and happy to start it.

**Comment by ehuss** [comment]
> the issue with json flag - this is my observation and i could be wrong - is that they are only available in the nightly builds. we cannot expect tool users to use only the nightly builds. 

If you are referring to the `--format=json` flag, it can be used on stable with the `-Zunstable-options` flag. There is a hole in the unstable restrictions for libtest, but not one anyone has been motivated to close.

We can also stabilize `--format=json` when used in conjunction with `--list` without stabilizing the JSON output as a whole.  (It's a bit unfortunate that every tool is using a different flag here, but I think it is not worth trying to address that.) I don't want to speak for the libs team, but I would expect that to be something that is easier to do.

---

I agree with @Manishearth that there probably should be more discussion. I assume there aren't already standard formats for just listing tests (such as junit)?  I would starting a thread on internals to see if anyone in a wider audience has any feedback for adding a JSON format for listing tests, along with a proposed structure.  I would also write a comment on #49359 soliciting feedback (since some people following that probably have some familiarity with machine-readable test integration).

**Comment by parthopdas** [comment]
@ehuss 

few things to set the context
1. typically test runners like in [vscode](https://marketplace.visualstudio.com/items?itemName=swellaby.vscode-rust-test-adapter) and soon in [visual studio](https://marketplace.visualstudio.com/items?itemName=kitamstudios.RustAnalyzer) work in 2 phases - test discovery then test execution. test discovery lists tests on a test explorer from which the dev can then run / debug tests.
1. i did explore the static analysis option that rust-analyzer on vscode uses to populate the code lens. however this is a no-go for me as visual studio does not do the same and would require a fair bit of effort + degraded user experience (2 x RLS per IDE) to implement.

point#1 means we cannot run the test and then populate the test explorer experience -> that beats the point of test explorer.

regarding the "--message-format json" it does not work for listing.
image

it works only for running
image

**Comment by parthopdas** [comment]
i am really thankful to the folks for jumping in with their comments / perspectives in <24hrs. i love it ❤️.

however i also feel a lack of direction of what needs to be done. there is a 'pressing business need' so to speak and *i do wish to do the right thing* but in phase wise and balanced manner - prefer not to boil the ocean first up and do so sequentially.

@petrochenkov @matklad @Manishearth @ehuss @m-ou-se based on your comments above i recommend the following approach
1. divide this issue into 2 phases (you can trust me to see both through to completion)
   1. unblock tooling requirement of listing tests (this change with perhaps modification of the format)
   1. get consensus and push through a proper fix through "--message-format json" for test listing as well => through RFC proces.
1. start discussions on internal thread for long and short term phases => this will result in a --message-format RFC i believe
1. solicit feedback on #49359 for long and short term phases
1. incorporate feedback into this PR for formatting / nits / code changes (esp. so we dont break any existing consumers) => one of you approve this PR so tooling is unbocked.
1. implement the RFC and therefore the long term "proper" solution.

wdyt?

**Comment by ehuss** [comment]
Yea, what I think I and others are suggesting here is to add a JSON format for listing tests (instead of extending the textual format proposed here).  Per the comments at https://github.com/rust-lang/rust/pull/107785#issuecomment-1422474199, I don't think we can change the existing output format, which I would agree with.

I would not expect there to be significant procedure for just introducing a JSON format for listing tests, the design space is not that large. Just take a little bit of time to see if anyone has any insight to contribute.

**Comment by parthopdas** [comment]
@ehuss may be i misunderstood. 

are you suggesting the following poa?
1. keep console listing as is
1. extend "--message-format json" for use with --list (i.e. without runnng)
1. solicit feedback internals + #107785 to ensure we're not missing anything obvious?

if so that is simpler and i am happy to start on it.

@petrochenkov @matklad @Manishearth @m-ou-se let us know if you have any comments.

**Comment by Manishearth** [comment]
> however i also feel a lack of direction of what needs to be done. there is a 'pressing business need' so to speak and **i do wish to do the right thing** but in phase wise and balanced manner - prefer not to boil the ocean first up & sequentially.

I think you have mismatched expectations as to the speed at which things move here.

Any feature addition, even if it's just a flag, is going to need to be properly thought through, because it's going to be supported _forever_. This is not something we take lightly. I think it's very unlikely that this PR will be merged without significant further discussion. We don't need to boil the ocean, but we _do_ need to make sure we understand the design space and plan accordingly.

This applies doubly so given that the test runner is not _quite_ owned by any team that can take responsibility here.

I don't see a "pressing business need at all". I think this is an interesting feature to have, but by no means super urgent.


Also, I recommend checking out [cargo nextest](https://github.com/nextest-rs/nextest); an unofficial test runner, which may be more amenable to quick feature requests if you have a pressing need.

**Comment by parthopdas** [comment]
ok @Manishearth how do you recommend i proceed? like i said i want to do the right thing. 

also i am happy for it to take as long as it does.

is this plan good enough for you?

1. keep console listing as is (revert the console.rs changes)
1. extend "--message-format json" for use with --list (i.e. without runnng)
1. solicit feedback internals + https://github.com/rust-lang/rust/issues/49359 to ensure we're not missing anything obvious?

regarding the other points nextest doesn't work for me because it requires users to add another package - rust-analyzer.vs should work out of the box i.e. 0 friction.

for now we can keep aside whether this is a critical or a good to have - that depends on how you see things.

**Comment by Manishearth** [comment]
I think for now just drop any code changes, and work on soliciting feedback and coming up with a holistic plan for all of the things involved. Make an internals post and then an RFC based on the feedback you get there (and the feedback you've gotten here so far)

**Comment by parthopdas** [comment]
@Manishearth okay sure.

so the new plan is:
1. can you point me to the 'internals forum"? is there an example / template post you'd like me to follow?
1. my initial post will start on the same lines i.e. business need, current inadequacy, recommendations to json format for listing and pointer this pr and #49359
1. give it a week, draft up a RFC (please point me to the process) based on the discussion
1. take it forward from there. 

okay with you?

**Comment by Manishearth** [comment]
@parthopdas internals.rust-lang.org

You could post a "pre-RFC" using the [RFC template](https://github.com/rust-lang/rfcs/blob/master/0000-template.md) but not necessarily fillign in _every_ section

But actually I suggest just making a post and talk about the motivation, and _solicit other motivations_ in the same area so you know what the set of needs are. And see what people propose as designs.

One concern I have is that you seem very very new to these processes, and it might be better if you can find someone more familiar who has time to help you shepherd this. I cannot.

> okay with you?

FWIW i'm not the person you need to convince, you need to convince the community and the teams :smile:

**Comment by parthopdas** [comment]
okay i'll start on it. thanks for the discussions everyone. love you guys! ❤️

@Manishearth let me know if you have anyone in mind who can shepherd this through.

**Comment by parthopdas** [comment]
@Manishearth please take a look https://internals.rust-lang.org/t/pre-rfc-enhancing-cargo-test-list-for-use-by-ide-test-explorers-runners/18308

**Comment by Manishearth** [comment]
> regarding the other points nextest doesn't work for me because it requires users to add another package - [rust-analyzer.vs](https://marketplace.visualstudio.com/items?itemName=kitamstudios.RustAnalyzer) or any other tool for that matter, should work out of the box i.e. 0 friction.

Actually, this doesn't apply: the entire purpose of this feature is to work with _other third party packages_ (this wasn't clear to me earlier). Those packages have other methods available to them for doing test discovery: looking at the code in the file, or using runners like nextest. They can choose to include a nextest binary if they want.

**Comment by Manishearth** [comment]
Let me put my comment there

**Comment by Manishearth** [comment]
cc @calebcartwright : You seem to maintain vscode-rust-test-adapter, do you think you can help here?

**Comment by Manishearth** [comment]
Also cc @matklad who probably has a better idea of IDE integration stuff

**Comment by calebcartwright** [comment]
> cc @calebcartwright : You seem to maintain vscode-rust-test-adapter, do you think you can help here?

I confess I'm a bit I'm unsure whether we're supposed to continue the dialog here or shift to internals, but will respond here for now.

From my pov I think it's fair to say that there's a lot that's already _technically_ possible with the existing `cargo test ...` surface as evidenced by existing editor/ide tooling. However, it's absolutely true that it's rather painful and brittle for those integrations to get the requisite data, and I did have to rely on some kludges even for the relatively minimal set of capabilities provided by the vscode-rust-test-adapter extension.

I'd personally love to see some libtest improvements in output support to make test-related ide/editor features more feasible, and I do think those types of features can play a non-insignificant factor in the overall developer experience with a language.

I'm not definitively sure who owns libtest either ([sounds like there may have been plans to establish a test team?](https://github.com/rust-lang/rust/issues/49359#issuecomment-1173034401)). I could probably make time to help assist an effort to make improvements, though I wouldn't have the bandwidth to fully drive the effort either.

**Comment by parthopdas** [comment]
@Manishearth what does it entail to take up 'ownership' of the testlib aspects? 

i am happy to contribute to that and other areas depending on my needs but what in addition to that?

**Comment by Manishearth** [comment]
No, it's about _team_ ownership.

Right now I think it mostly falls between libs and devtools. The creation of a formalized devtools subteam would be nice for it, but it would need to be run or at least advised by someone familiar with how the Rust project works. If you can find someone willing to take that on, we can try something.

In general libtest has been considered "done" (or, low priority) for a while which is why there hasn't ever been much motion on it. There's the custom test frameworks RFC ([tracking issue](https://github.com/rust-lang/rust/issues/50297)) that didn't ever get driven further and that would be a nice task for a team picking this up, but kinda big.

**Comment by parthopdas** [comment]
since we are on the topic - i'd state my opinion as well on custom test frameworks [RFC 50297](https://github.com/rust-lang/rust/issues/50297) and a proposal for it [jrenner proposal](https://www.jrenner.net/rust/testing/2018/08/06/custom-test-framework-prop.html)

i like the current approach better: doc / bench / tests with just 1 attribute [#test] with ability to extend e.g. test-case / quickcheck crates. even there while doc is an innovation but bench is also probably not required. [RFC 50297](https://github.com/rust-lang/rust/issues/50297) imho is overcomplicating testability aspects of rust code.

> We could stabilize #[bench] and extend libtest with setup/teardown and other requested features.

i think this is the way to go - which is what we appear to be doing.

curiously the RFC is missing a prior art section. and while other ecosystems do have custom test frameworks (e.g. .net world has the xunit family) it is not to the extent the rfc is proposing. 

imho end goal of testing is to rapidly provide feedback on the architecture itself. harder something is to test, lesser you can prove its correctness and more the complexity will seep into the end user experience.

i am a believer of 1000s of fine grained unit tests, 100ish of integration tests and 10s of end to end Acceptance tests. i cover that philosophy in bunch of talks at [YT @parthopdas](https://www.youtube.com/@parthopdas). simpler is better - lesser the constraints, the more degrees of freedom, the more scenarios can be covered (e.g. if just one test provides which in itself is complex, doesn't meet my needs what then? in addition more complex you make a test framework (e.g. setup teardown), the more complex testing itself becomes and lesser of it we do.

what i would love to see (and willing to make happen) is investment in tools that make testing experience sexy & a breeze - for the very same reasons.

case in point: [ncrunch](https://www.ncrunch.net/) but for rust. and i believe with just minor tweaks to the existing infra, we have everything we need from the current testlib implementation.

**Pull request closed without merging**